File: auth-backend/core/domain/auth/permission.py
"""
Permission Domain Models
Fine-grained resource-level permissions
"""
from dataclasses import dataclass
from typing import Optional
from datetime import datetime
from enum import Enum


class PermissionAction(str, Enum):
    """Permission actions"""
    CREATE = "create"
    READ = "read"
    UPDATE = "update"
    DELETE = "delete"
    MANAGE = "manage"  # Full control


# ResourceType is now a FREE STRING!
# No enum needed - allows adding new resource types without code changes.
# Examples: "ticket", "project", "order", "invoice", "deal", "document", etc.
# Each system/client can define their own resource types dynamically.


@dataclass
class Permission:
    """
    Fine-grained permission model.
    
    Allows resource-level access control beyond basic RBAC.
    
    **Design Note**: resource_type is a FREE STRING (not enum) for maximum
    flexibility. Each system can define its own resource types without
    modifying the auth system code.
    """
    id: Optional[str] = None
    user_id: str = ""
    client_id: str = ""  # Multi-tenant
    resource_type: str = ""  # String livre! Pode ser "ticket", "invoice", "project", etc.
    resource_id: Optional[str] = None  # None = all resources of this type
    action: Optional[PermissionAction] = None
    granted_by: Optional[str] = None  # User who granted permission
    created_at: Optional[datetime] = None
    
    def __post_init__(self):
        if self.created_at is None:
            self.created_at = datetime.utcnow()
        # Normalize resource_type
        if self.resource_type:
            self.resource_type = self.resource_type.lower().strip()
    
    def allows(self, action: PermissionAction, resource_id: Optional[str] = None) -> bool:
        """Check if permission allows action on resource"""
        # MANAGE allows everything
        if self.action == PermissionAction.MANAGE:
            return True
        
        # Check action match
        if self.action != action:
            return False
        
        # Check resource (if permission is resource-specific)
        if self.resource_id and resource_id and self.resource_id != resource_id:
            return False
        
        return True

File: auth-frontend/src/core/domain/permission.ts
/**
 * Permission Domain Models
 * Fine-grained resource-level permissions
 */

export type PermissionAction = 'create' | 'read' | 'update' | 'delete' | 'manage';

export interface Permission {
  id: string;
  user_id: string;
  client_id: string;
  resource_type: string; // String livre - permite flexibilidade máxima
  resource_id?: string; // undefined = todas os recursos deste tipo
  action: PermissionAction;
  granted_by?: string;
  created_at: Date;
}

export interface GrantPermissionRequest {
  user_id: string;
  resource_type: string;
  action: PermissionAction;
  resource_id?: string;
}

export interface PermissionListResponse {
  permissions: Permission[];
  total: number;
}

/**
 * Helper class for permission checking
 */
export class PermissionChecker {
  /**
   * Check if a permission allows a specific action on a resource
   */
  static allows(
    permission: Permission,
    action: PermissionAction,
    resourceId?: string
  ): boolean {
    // MANAGE allows everything
    if (permission.action === 'manage') {
      return true;
    }

    // Check action match
    if (permission.action !== action) {
      return false;
    }

    // Check resource (if permission is resource-specific)
    if (permission.resource_id && resourceId && permission.resource_id !== resourceId) {
      return false;
    }

    return true;
  }

  /**
   * Check if user has permission for action on resource
   */
  static hasPermission(
    permissions: Permission[],
    resourceType: string,
    action: PermissionAction,
    resourceId?: string
  ): boolean {
    return permissions.some(
      (p) =>
        p.resource_type === resourceType &&
        this.allows(p, action, resourceId)
    );
  }
}

